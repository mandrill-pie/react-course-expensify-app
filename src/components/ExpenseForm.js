import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
	// need a constructor to get access to props
	constructor(props) {
		super(props);

		this.state = {
			description: props.expense ? props.expense.description : '',
			amount: props.expense ? props.expense.amount : '',
			createdAt: props.expense ? moment(props.expense.createdAt): moment(),
			note: props.expense ? props.expense.note : '',
			calendarFocused: false,
			error: false,
		};
	}

	onDescriptionChange = (e) => {
		const description = e.target.value;
		this.setState(() => ({ description }));
	}
	onNoteChange = (e) => {
		const note = e.target.value;
		this.setState(() => ({ note }));
	}
	onAmountChange = (e) => {
		const amount = e.target.value;
		if (!amount || amount.match(/^\d{1,10}(\.\d{0,2})?$/)) {
			this.setState(() => ({ amount }));
		}
	}
	onDateChange = (createdAt) => {
		if (createdAt) {
			this.setState(() => ({ createdAt }));
		}
	}
	onFocusChange = ({ focused }) => {
		this.setState(() => ({ calendarFocused: focused }));	
	}
	onSubmit = (e) => {
		e.preventDefault();

		if (!this.state.description || !this.state.amount) {
			this.setState(() => ({ error: 'Please provide a description and amount!' }));
		} else {
			this.setState(() => ({ error: false }));
			// call props.onSubmit() that gets passed to this component
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount),
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note,
			});
		}

	}
	render() {
		return (
			<form className="form" onSubmit={this.onSubmit}>
				{
					this.state.error &&
					<p className="form__error">{this.state.error}</p>
				}
				<input
					className="input-text"
					type="text"
					placeholder="Description"
					autoFocus
					value={this.state.description}
					onChange={this.onDescriptionChange} />
				<input
					className="input-text"
					type="text" // number
					placeholder="Amount"
					value={this.state.amount}
					onChange={this.onAmountChange} />
				<SingleDatePicker
					date={this.state.createdAt}
					onDateChange={this.onDateChange}
					focused={this.state.calendarFocused}
					onFocusChange={this.onFocusChange}
					numberOfMonths={1}
					isOutsideRange={() => false} />
				<textarea
					className="textarea"
					placeholder="Add a note"
					onChange={this.onNoteChange}
					value={this.state.note}>
				</textarea>

				<input className="btn" type="submit" value="Save expense" />
			</form>
		);
	}
}