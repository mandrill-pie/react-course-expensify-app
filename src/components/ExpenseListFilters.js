import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setFilterText, sort, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			calendarFocused: null,
		};
	}

	onDatesChange = ({ startDate, endDate }) => {
		this.props.setStartDate(startDate);
		this.props.setEndDate(endDate);
	}

	onFocusChange = (calendarFocused) => {
		this.setState(() => ({ calendarFocused }));
	}

	onTextChange = (e) => {
		this.props.setFilterText(e.target.value);
	}

	onSortChange = (e) => {
		this.props.sort(e.target.value);
	}

	render = () => (
		<div className="content-container">
			<div className="input-group">
				<div className="input-group__item">
					<input
						className="input-text"
						type="text"
						placeholder="Search expenses"
						value={this.props.filters.text}
						onChange={this.onTextChange} />
				</div>
				<div className="input-group__item">
					<select
						className="select"
						value={this.props.filters.sortBy}
						onChange={this.onSortChange}>
						<option value="date">Date</option>
						<option value="amount">Amount</option>
					</select>
				</div>
				<div className="input-group__item">
					<DateRangePicker
						startDate={this.props.filters.startDate}
						endDate={this.props.filters.endDate}
						onDatesChange={this.onDatesChange}
						focusedInput={this.state.calendarFocused}
						onFocusChange={this.onFocusChange}
						showClearDates={true}
						numberOfMonths={1}
						isOutsideRange={() => false} />
				</div>
			</div>
		</div>
	);
}

const mapStateToProps = (state) => ({
	filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
	setStartDate: (startDate) => dispatch(setStartDate(startDate)),
	setEndDate: (endDate) => dispatch(setEndDate(endDate)),
	setFilterText: (text) => dispatch(setFilterText(text)),
	sort: (value) => dispatch(sort(value)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);