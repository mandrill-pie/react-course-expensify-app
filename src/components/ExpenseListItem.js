import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({id, description, amount, createdAt}) => (
	<Link className="list-item" to={`/edit/${id}`}>
		<div>
			<h3 className="list-item__title">{description}</h3>
			<span className="list-item__date">{moment(createdAt).format('YYYY-MM-DD')}</span>
		</div>
		<div>
			<h3 className="list-item__amount">
				{numeral(amount).format('$0,0.00')}
			</h3>
		</div>
	</Link>
);

export default ExpenseListItem;