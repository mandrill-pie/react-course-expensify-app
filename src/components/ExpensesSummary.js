import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import { Link } from 'react-router-dom';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = (props) => {
	const expenseWord = props.expensesCount === 1 ? 'expense' : 'expenses';
	const formattedExpensesCount = numeral(props.expensesTotal).format('$0,0.00');
	return (
		<div className="page-header">
			<div className="content-container">
				<h2 className="page-header__title">
					Viewing <span>{props.expensesCount}</span> {expenseWord} totalling <span>{formattedExpensesCount}</span>
				</h2>
				<div className="page-header__actions">
					<Link className="btn" to="/create">Add expense</Link>
				</div>
			</div>
		</div>
	);
};

const mapStateToProps = (state) => {
	const visibleExpenses = selectExpenses(state.expenses, state.filters);
	return {
		expensesCount: visibleExpenses.length,
		expensesTotal: selectExpensesTotal(visibleExpenses)
	};
};

export default connect(mapStateToProps)(ExpensesSummary);