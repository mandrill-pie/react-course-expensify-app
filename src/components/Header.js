import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';
import selectExpenses from '../selectors/expenses';

export const Header = ({ user, startLogout }) => (
	<header className="header">
		<div className="content-container">
			<div className="header__content">
				<Link className="header__title" to="/dashboard">
					<h1>Expensify</h1>
				</Link>
				<button className="btn btn--link" onClick={startLogout}>Logout {user.displayName}</button>
			</div>
		</div>
	</header>
);

const mapStateToProps = (state) => ({
	user: state.auth.user
});

const mapDispatchToProps = (dispatch) => ({
	startLogout: () => dispatch(startLogout())
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);