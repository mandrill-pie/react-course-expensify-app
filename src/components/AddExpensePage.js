import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
	// no constructor needed to get access to props!

	onSubmit = (expense) => {
		// dispatch by using addExpense (defined in mapDispatchToProps)
		// workaround for testing, used instead of:
		// props.dispatch(addExpense(expense));
		this.props.addExpense(expense);
		// redirect
		this.props.history.push('/');
	}

	render() {
		return (
			<div>
				<ExpenseForm
					onSubmit={this.onSubmit}
				/>
			</div>
		);
	}
}

const mapDispatchToProps = (dispatch) => ({
	// Give AddExpensePage component access to addExpense()
	addExpense: (expense) => dispatch(addExpense(expense))
});

// give AddExpensePage access to props
// so it can use props.dispatch()
// do not pass a mapStateToProps() function
// but do pass a mapDispatchToProps() (used in testing)
export default connect(false, mapDispatchToProps)(AddExpensePage);