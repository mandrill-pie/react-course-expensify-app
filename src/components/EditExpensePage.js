import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {

	onSubmit = (updates) => {
		// dispatch
		this.props.startEditExpense(this.props.expense.id, updates);
		// redirect
		this.props.history.push('/');
	}

	onRemove = () => {
		this.props.startRemoveExpense({ id: this.props.expense.id });
		this.props.history.push('/');
	}

	render() {
		return (
			<div>
				<div className="page-header">
					<div className="content-container">
						<h2 className="page-header__title">Edit Expense</h2>
					</div>
				</div>
				<div className="content-container">
					<ExpenseForm
						expense={this.props.expense}
						onSubmit={this.onSubmit}
					/>
					<button className="btn btn--secondary" onClick={this.onRemove}>Remove Expense</button>
				</div>
			</div>
		);
	}
}

const mapStateToProps = (state, props) => ({
	// populate props.expense with the required expense (based on id)
	expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
	// Give parent component (EditExpensePage) access to editExpense()
	// Needs access to props (second argument)! <- WHY?
	startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
	startRemoveExpense: (data) => dispatch(startRemoveExpense(data)),
});

// give EditExpensePage access to props
// so it can use props.dispatch()
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);