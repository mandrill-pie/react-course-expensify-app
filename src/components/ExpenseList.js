import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export class ExpenseList extends React.Component {
	constructor(props) {
		super(props);
	}

	render() {
		return (
			<div className="content-container">
				<div className="list-header">
					<div className="show-on-mobile">Expenses</div>
					<div className="show-on-desktop">Expense</div>
					<div className="show-on-desktop">Amount</div>
				</div>
				<div className="list-body">
					{
						this.props.expenses.length === 0 ? (
							<div className="list-item list-item--message">
								<span>No expenses found</span>
							</div>
						) : (
							<div>
								{
									this.props.expenses.map((expense) => (
										<ExpenseListItem key={expense.id} {...expense} />
									))
								}
							</div>
						)
					}
				</div>
			</div>
		);
	}
}

const mapStateToProps = (state) => ({
	expenses: selectExpenses(state.expenses, state.filters),
});

export default connect(mapStateToProps)(ExpenseList);