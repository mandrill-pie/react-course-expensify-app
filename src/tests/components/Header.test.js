import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../components/Header';

// run before running each test
let startLogout, wrapper;

beforeEach(() => {
	// create spy functions
	startLogout = jest.fn();
	wrapper = shallow(
		<Header
			startLogout={startLogout} />
	);
});

test('render header', () => {
	expect(wrapper).toMatchSnapshot();
});

test('should call startLogout on button click', () => {
	wrapper.find('button').simulate('click');
	expect(startLogout).toHaveBeenCalled();
});

