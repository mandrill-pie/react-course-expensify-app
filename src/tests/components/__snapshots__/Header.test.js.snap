// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render header 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        replace={false}
        to="/dashboard"
      >
        <h1>
          Expensify
        </h1>
      </Link>
      <button
        className="btn btn--link"
        onClick={[Function]}
      >
        Logout 
        Logged In User
      </button>
    </div>
  </div>
</header>
`;
